Add localization to Wireguard service tray menu

spylogsster opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/30959

- Renamed `components\resources\strings\brave_components_resources_XX.xtb` - > `components\resources\strings\brave_components_strings_XX.xtb` 
- Move Wireguard service strings out of `brave_vpn_strings.grdp` to `components\resources\brave_components_strings.grd` because `create_string_rc.py` doesnt work with grdp.


<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-arm64, CI/skip-linux-x64, CI/skip-android, CI/skip-macos, CI/skip-ios, CI/skip-windows-arm64, CI/skip-windows-x64, CI/skip-windows-x86 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- Check the tray menu has translated items for different locales

simonhong commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/resources/BUILD.gn:
```diff
@@ -3,9 +3,35 @@
+generate_embedded_i18n("generate_strings") {
+  visibility = [ ":strings" ]
+  extractor_datafile = "//brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/resources/create_wireguard_string_rc.py"
+  grdfile_folder = "//chrome/app/"
```
it seems layer violation?

spylogsster replied:
This is folder where it will take grd file to extracat strings and create a header in gen. The header created inside current layer  so should be ok. Moving this code to `chrome` will produce the header inside chrome and we will not be able to use it inside the component. cc @brave/string-reviewers-team

simonhong replied:
Got it.
Another question - In this PR, new strings are added into `brave/app/brave_strings.grd`.
Curious they could be added into grd files in `brave/component/` instead?

spylogsster replied:
when I try to do this there's next error:
```
[`brave_components_resources_af.xtb`](ninja: error: '../../brave/components/resources/strings/brave_components_resources_af.xtb', needed by 'gen/brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/resources/wireguard_strings.h', missing and no known rule to make it)
```

if I created that file manually then got same error for `brave_components_resources_as.xtb`. Looks like we do not support all languages, maybe @mkarolin can advise something how to fix it or keep current way.

mkarolin replied:
We can't add new strings directly into `brave/app/brave_strings.grd`. This file is auto updated from `src/chrome/app/chromium_strings.grd` (+2 entries we add in script), so your changes would be removed when we run rebase the next time.

`brave_component_resources` seems like a good option, but as you noticed we don't support all the languages that Chromium supports (see https://github.com/brave/brave-browser/wiki/Strings-and-Localization#information-for-submitting-localization-orders section - the languages we support are bolded).

spylogsster replied:
Thanks @mkarolin , updated code to exclude unused locales from component grds, but there are some errors:

If I use `brave_component_resources` throws:
```
There were files that exist on disk but were not found in the .grd file '../../brave/components/resources/brave_components_resources.grd':
../../brave/components/resources/strings/brave_components_resources_am.xtb
../../brave/components/resources/strings/brave_components_resources_ar.xtb
../../brave/components/resources/strings/brave_components_resources_bg.xtb
....
../../brave/components/resources/strings/brave_components_resources_zh-TW.xtb
[126/31408] CXX obj/mojo/core/ports/ports/node.obj
```

If I use `brave_component_strings` throws: 
```
ninja: error: '../../brave/components/resources/strings/brave_components_strings_am.xtb', needed by 'gen/brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/resources/wireguard_strings.h', missing and no known rule to make it
ninja: Entering directory `D:\repos\brave\brave-browser\src\out\Component'
```

spylogsster replied:
Finally it required 2 things to make it worked:
1) Rename `components\resources\strings\brave_components_resources_XX.xtb` - > `components\resources\strings\brave_components_strings_XX.xtb` 
2) Move Wireguard service strings out of `brave_vpn_strings.grdp` to `components\resources\brave_components_strings.grd` because `create_string_rc.py` doesnt work with grdp.

It looks working now, please take a look @mkarolin

mkarolin replied:
This looks good to me @spylogsster. Could you, please, also update the comment here https://github.com/brave/brave-core/blob/master/build/commands/lib/l10nUtil.js#L57.

spylogsster replied:
вщту

simonhong commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/l10n_util.cc:
```diff
@@ -0,0 +1,59 @@
+
+}  // namespace
+
+std::u16string GetLocalizedString(UINT base_message_id) {
```
Sorry, I don't understand why this api is needed. Can't we use `GetStringUTF16()` here?

spylogsster replied:
GetStringUTF16 works over `ui::ResourceBundle` which is used only inside chrome. All `generate_embedded_i18n` usages also have similar implementations:
- updater  https://source.chromium.org/chromium/chromium/src/+/main:chrome/updater/win/ui/l10n_util.h;l=6?q=l10n_&sq=&ss=chromium
- installer https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/util/l10n_string_util.h;l=14?q=l10n_string_util.h&sq=&ss=chromium

I also tried to initialize resource bundle but it did work as expected.

simonhong replied:
Found related docs - https://source.chromium.org/chromium/chromium/src/+/main:docs/updater/design_doc.md;l=272-288

simonhong reviewed (approved):
++

mkarolin reviewed (approved):
LGTM with a comment to fix https://github.com/brave/brave-core/blob/master/build/commands/lib/l10nUtil.js#L57

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/l10n_util.cc:
```diff
@@ -0,0 +1,59 @@
+
+}  // namespace
+
+std::u16string GetLocalizedString(UINT base_message_id) {
```
I don't understand what's going on here, the chromium code already takes care of this. This should be added to a pak file just like in other cases and use `ui::ResourceBundle::GetSharedInstance()`

bridiver replied:
the point is that this executable is installed with Brave so it can just load the pak files that come with Brave using `ui::ResourceBundle`

bridiver replied:
This is different from the installer where Brave may or may not already exist when it runs

spylogsster replied:
I tried to to this and it doesnt work, the resources seems stripped.
```
[0707/222952.392:FATAL:resource_bundle.cc(1199)] Check failed: !data.empty(). Unable to find resource: 54404. If this happens in a browser test running on Windows, it may be that dead-code elimination stripped out the code that uses the resource, causing the resource to be stripped out because the resource is not used by chrome.dll. See https://crbug.com/1181150.
```
Also I do not see similar usages of `InitSharedInstanceWithPakPath`, it is used mostly in tests or to load test paks from examples ([link](https://source.chromium.org/search?q=InitSharedInstanceWithPakPath&ss=chromium))

Opened separate PR for this approach to play if need https://github.com/brave/brave-core/pull/19202/files, it fails in runtime because no strings found in the pak file. 

Additionally worth to mention that using runtime pak files we implicitly dependent on external resource. Explicit better than implicit and Chromium recommended way to do compile time localization for separate executables is using `generate_embedded_i18n` which `defines a rule to generate localized string .rc and .h files that can be embedded directly into a library or executable if desired` ([link](https://source.chromium.org/chromium/chromium/src/+/main:base/win/embedded_i18n/generate_embedded_i18n.gni;l=8?q=generate_embedded_i18n&ss=chromium)) and used in others Chromium executables ([credential_provider](https://source.chromium.org/chromium/chromium/src/+/main:chrome/credential_provider/gaiacp/gcp_utils.cc;l=923?q=gcp_utils.cc&ss=chromium%2Fchromium%2Fsrc:chrome%2Fcredential_provider%2F), [updater](https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:chrome/updater/win/ui/l10n_util.cc;l=47), [installer](https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/util/l10n_string_util.cc;l=66?q=GetLocalizedString&ss=chromium)).

bridiver commented on components/resources/brave_components_strings.grd:
```diff
@@ -105,60 +105,60 @@
-    <file path="strings/brave_components_resources_vi.xtb" lang="vi" />
-    <file path="strings/brave_components_resources_zh-CN.xtb" lang="zh-CN" />
-    <file path="strings/brave_components_resources_zh-TW.xtb" lang="zh-TW" />
+    <file path="strings/brave_components_strings_am.xtb" lang="am" />
```
why  are you renaming all of these files? While they probably should be `*_strings` instead of `*_resources`, I think that is just confusing what is going on in this PR and it should be separate

spylogsster replied:
generate_embedded_i18n expects that xtb files have same name prefix as initial .grd file. https://source.chromium.org/chromium/chromium/src/+/main:base/win/embedded_i18n/generate_embedded_i18n.gni;l=129?q=generate_embedded_i18n&ss=chromium%2Fchromium%2Fsrc 
This is also visible by others grd/xtb files in the Chromium repo

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/resources/BUILD.gn:
```diff
@@ -3,9 +3,69 @@
+
+  # Exclude languages not supported by Brave
+  # https://github.com/brave/brave-browser/wiki/Strings-and-Localization#information-for-submitting-localization-orders
+  brave_supported_i18_locales = default_embedded_i18_locales - [
```
This is not maintainable. If the list of languages changes, this will be out-of-date

spylogsster replied:
I think the list of unmaintainable locales should be moved to the config.gni to use from one place, @mkarolin wdyt?

bridiver commented on components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/l10n_util.cc:
```diff
@@ -0,0 +1,59 @@
+#include "base/win/atl.h"
+#include "base/win/embedded_i18n/language_selector.h"
+#include "base/win/i18n.h"
+#include "brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/status_tray/resources/wireguard_strings.h"
```
why are `wireguard_strings` under `status_tray`?

spylogsster replied:
renamed

spylogsster commented:
Since we are reading strings from paks we have to be dependent on repack target for the browser to regenerate required resources for the service. Closing this PR because it is impossible to do from //components/ and opened for review anptehr one
https://github.com/brave/brave-core/pull/19204 where I moved service to brave/browser/ and added dependency to ` "//chrome:packed_resources",`
